pytissueoptics/rayscattering/tests/opencl/src/testGeneralOpenCL.py

```diff
@@ -0,0 +1,544 @@
+
+        knl = self.program.test_local_id # copy local_id into array
+
+        knl(queue, (nWorkUnits,), None, valueBuffer.data)
+
+        for i, value in enumerate(valueBuffer):
+            self.assertEqual(i%32, value)
+
+    def test_trivial_kernel_local_id_set_workgroup_size(self):
+        """
+        The number of groups must divide nWorkUnits evenly
+        """
+
+        queue = pycl.CommandQueue(self.context)
+        nWorkUnits = 100
+        valueBuffer = clArray(cq=queue, shape=(nWorkUnits,), dtype=cl.cltypes.uint)
+
+        knl = self.program.test_local_id # copy local_id into array
+
+        for i in range(100):
+            try:
+                knl(queue, (nWorkUnits,), (i,), valueBuffer.data)
+                self.assertEqual(nWorkUnits%i, 0)
+
+                for j, value in enumerate(valueBuffer):
+                    self.assertEqual(j%nWorkUnits, value)
+            except:
+                pass
+
+
+    # @unittest.skip('For information only')
+    # def test_extract_many_ids(self):
+    #     nWorkUnits = 100
+    #     valueBuffer = Buffer(nWorkUnits, value=0, dtype=cl.cltypes.uint)
+
+    #     source_path = os.path.join(OPENCL_SOURCE_DIR, "test_opencl.c")
+    #     program = CLProgram(source_path)
+    #     program.launchKernel(kernelName='test_extract_many_ids', N=nWorkUnits, arguments = [valueBuffer])
+    #     for i, value in enumerate(valueBuffer.hostBuffer):
+    #         print(i,value)
+
+    def test_compute_global_id_from_local_id_uniform_sizes(self):
+        queue = pycl.CommandQueue(self.context)
+        nWorkUnits = 128 # Multiplie of 32
+        valueBuffer = clArray(cq=queue, shape=(nWorkUnits,), dtype=cl.cltypes.uint)
+
+        knl = self.program.test_compute_global_id_from_local_id # copy local_id into array
+        knl(queue, (nWorkUnits,), None, valueBuffer.data)
+
+        for i, value in enumerate(valueBuffer):
+            self.assertEqual(value, i)
+
+    @unittest.expectedFailure
+    def test_compute_global_id_from_local_id_NON_uniform_sizes(self):
+        """
+        My assumption was that workunits would be fed in order but that is not true.
+        I don't know how to compute the global_id from the local parameters then
+        It does not matter: use get_global_id() but this is a basic issue in 
+        understanding OpenCL for me.
+        """
+
+        queue = pycl.CommandQueue(self.context)
+        nWorkUnits = 100 # NOT A MULTIPLE of 32
+        valueBuffer = clArray(cq=queue, shape=(nWorkUnits,), dtype=cl.cltypes.uint)
+        local_sizes = clArray(cq=queue, shape=(nWorkUnits,), dtype=cl.cltypes.uint)
+
+        knl = self.program.test_compute_global_id_from_local_id_nonuniform # copy local_id into array
+        knl(queue, (nWorkUnits,), None, valueBuffer.data, local_sizes.data)
+
+        for i, value in enumerate(valueBuffer):
+            self.assertEqual(value, i)
+
+        self.assertEqual(np.sum(local_sizes.hostBuffer), nWorkUnits)
```

Comment:
I run into a few issues with these tests on windows and linux with OpenCL 3.0. Same results on an old python 3.8 env and fresh 3.10. 

- Line 491: `AttributeError: 'Array' object has no attribute 'hostBuffer'`
- Line 425:
```
   -->  self.assertEqual(i%32, value)
AssertionError: 0 != cl.Array(32, dtype=uint32)
```

- And on windows only:
Line 429: test_trivial_kernel_local_id_set_workgroup_size : `exit code 0xC0000094 (Integer division by zero)`